lib/openstack_query/query_wrapper.py

```diff
@@ -0,0 +1,353 @@
+from enum import Enum
+from typing import Tuple, Callable, Any, Dict, Optional, List, Union
+from openstack_api.openstack_wrapper_base import OpenstackWrapperBase
+from openstack_api.openstack_connection import OpenstackConnection
+
+from openstack_query.utils import (
+    check_filter_func,
+    check_kwarg_mapping,
+    prop_equal_to,
+    prop_not_equal_to,
+    prop_greater_than,
+    prop_less_than,
+    prop_younger_than,
+    prop_older_than,
+    prop_older_than_or_equal_to,
+    prop_younger_than_or_equal_to,
+    prop_any_in,
+    prop_not_any_in,
+    prop_matches_regex,
+)
+
+
+from exceptions.parse_query_error import ParseQueryError
+from exceptions.query_mapping_error import QueryMappingError
+
+from enums.query.query_presets import QueryPresets
+from tabulate import tabulate
+
+
+class QueryWrapper(OpenstackWrapperBase):
+    _PROPERTY_MAPPINGS = {}
+    _KWARG_MAPPINGS = {}
+    _NON_DEFAULT_FILTER_FUNCTION_MAPPINGS = {}
+    _DEFAULT_FILTER_FUNCTION_MAPPINGS = {}
+    _DEFAULT_FILTER_FUNCTIONS = {
+        QueryPresets.EQUAL_TO: prop_equal_to,
+        QueryPresets.NOT_EQUAL_TO: prop_not_equal_to,
+        QueryPresets.GREATER_THAN: prop_greater_than,
+        QueryPresets.LESS_THAN: prop_less_than,
+        QueryPresets.YOUNGER_THAN: prop_younger_than,
+        QueryPresets.YOUNGER_THAN_OR_EQUAL_TO: prop_younger_than_or_equal_to,
+        QueryPresets.OLDER_THAN_OR_EQUAL_TO: prop_older_than_or_equal_to,
+        QueryPresets.OLDER_THAN: prop_older_than,
+        QueryPresets.ANY_IN: prop_any_in,
+        QueryPresets.NOT_ANY_IN: prop_not_any_in,
+        QueryPresets.MATCHES_REGEX: prop_matches_regex,
+    }
```

Comment:
I was wondering about such a hierarchy based on the above:

```python
class QueryPropBase(abc.ABC):
   @abstractmethod
   def is_supported_query(preset: QueryPresets)

class QueryPropComparison(QueryPropBase):
    __FILTER_FUNCTIONS = {
        QueryPresets.EQUAL_TO: prop_equal_to,
        QueryPresets.NOT_EQUAL_TO: prop_not_equal_to,
        # ...
    }

   def is_supported_query(preset: QueryPresets):
       return preset in _FILTER_FUNCTIONS.keys()
```

Then this class can simply do:
```
def __init__(...):
  self._prop_handlers: List[QueryBase] = [QueryPropDate(), QueryPropComparisons()]

def get_handler(query: QueryPreset) -> QueryBase:
   for i in self._prop_handlers:
       if i.is_supported_query(query):
          return i
```

Then the rest of your code can simply duck type, since it knows that prop class has to support it's query

This will massively curtail / reduce the size of this file and it's responsibilities as it's already verging into a god class

Reply:
this makes sense thanks :+1:

Reply:
in your example we could put `prop_equal_to` and `prop_not_equal_to` as static methods in `QueryPropComparison` class